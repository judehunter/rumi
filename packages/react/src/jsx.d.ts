import {createRumi} from '@rumi/core/src/createRumi';

type CSS = Parameters<ReturnType<typeof createRumi>['css']>[0];

declare module 'react' {
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  interface DOMAttributes<T> {
    css?: CSS;
  }
  interface HTMLAttributes<T> extends AriaAttributes, DOMAttributes<T> {
    css?: CSS;
  }
}

declare global {
  namespace JSX {
    interface IntrinsicAttributes {
      css?: CSS;
    }
  }
}

export {};
