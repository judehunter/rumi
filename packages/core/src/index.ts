export {createRumi} from './createRumi';
export {flushCache} from './stylesheet';
