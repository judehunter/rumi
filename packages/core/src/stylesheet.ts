let styleTag: HTMLStyleElement | null = null;

const stylesheet: Record<string, {cssString: string; styles: any}> = {};

export const isInStylesheet = (className: string) => {
  return Object.keys(stylesheet).includes(className);
};

export const addToStylesheet = (
  className: string,
  cssString: string,
  styles?: any,
) => {
  stylesheet[className] = {cssString, styles};

  if (globalThis.document) {
    if (styleTag === null) {
      styleTag = globalThis.document.createElement('style');
      styleTag.setAttribute('data-rumi', '');

      globalThis.document.head.appendChild(styleTag);
    }

    styleTag.appendChild(globalThis.document.createTextNode(cssString));
  }
};
