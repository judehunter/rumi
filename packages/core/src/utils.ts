export type TOrArrayT<T> = T | T[];
