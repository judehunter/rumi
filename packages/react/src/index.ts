import {createRumi} from '@rumi/core';

console.log(createRumi);
