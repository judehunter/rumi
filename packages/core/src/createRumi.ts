import {stringify} from '@stitches/stringify';
import {toHash} from './hash';
import {all as mergeAll} from 'deepmerge';
import * as CSSType from 'csstype';
import {addToStylesheet, isInStylesheet} from './stylesheet';

type CSSBaseStyles = Pick<
  CSSType.Properties,
  'display' | 'color' | 'backgroundColor'
>;

type Contains<T extends string> =
  | `${T}${string}`
  | `${string}${T}${string}`
  | `${string}${T}`;

type UtilFunction = (value: any) => Partial<CSSBaseStyles>;

type RumiTheme = any;

type Media = Record<string, string>;

export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
};

export type RumiConfig = {
  utils: Record<string, UtilFunction>;
  theme?: RumiTheme;
  media?: Media;
  prefix?: string;
};

type RumiClassNames = string & {__rumi_brand__: true};

export const createRumi = <T extends RumiConfig>(cfg: T) => {
  type StylesObject =
    | (CSSBaseStyles & {
        [Prop in keyof T['media']]: StylesObject;
      } & {
        [K in Contains<'&'>]: StylesObject;
      })
    | {styles: StylesObject};

  type ShortCircuitStyles =
    | null
    | undefined
    | boolean
    | number
    | string
    | DeepPartial<StylesObject>;

  type TOrArrayT<T> = T | T[];

  let stylesheet = {} as Record<string, any>;

  const getCssText = ({reset = true} = {}) => {
    const ret = Object.values(stylesheet).join('');
    if (reset) stylesheet = {};
    return ret;
  };

  const transformSpecialProperties = (styles: any) => {
    const newObj = {} as Record<string, any>;

    for (const key of Object.keys(styles)) {
      // console.log('key', key);

      // const foundUtilFunction = Object.entries(cfg.utils).find(
      //   ([k]) => k === key,
      // )?.[1];

      // if (foundUtilFunction) {
      //   newObj = {...newObj, ...foundUtilFunction(styles[key])};
      //   continue;
      // }

      let foundAtRule = Object.entries(cfg.media ?? {}).find(
        ([k]) => k === key,
      )?.[1];

      if (foundAtRule) {
        foundAtRule = `@media ${foundAtRule}`;
        newObj[foundAtRule] = styles[key];
        if (newObj[foundAtRule] instanceof Object) {
          newObj[foundAtRule] = transformSpecialProperties(newObj[foundAtRule]);
        }
        continue;
      }

      if (styles[key] instanceof Object) {
        newObj[key] = transformSpecialProperties(styles[key]);

        continue;
      }

      newObj[key] = styles[key];
    }

    return newObj;
  };

  const mergeStyles = (styles: any[]) => {
    return mergeAll(styles);
  };

  const generateClassNames = (styles: any) => {
    const classes = {} as Record<string, string>;

    // an object which is a subobject of styles, without the atomic styles
    const whole = {} as any;
    for (const [k, v] of Object.entries(styles)) {
      // figure out if it's a special property or regular css
      if (
        ~k.indexOf('&') ||
        Object.entries(cfg.media ?? {}).find(([k2]) => k === k2)
      ) {
        // preserve whole styles
        whole[k] = v;
      } else {
        // break up into atomic styles
        // and stringify into distinct classes
        const atomicStyle = {
          [k]: v,
        };

        const className = `.${cfg.prefix ? cfg.prefix + '-' : ''}rumi-${toHash(
          atomicStyle,
        )}`;
        classes[className] = stringify({[className]: atomicStyle});
      }
    }

    // stringify whole styles as one
    if (Object.keys(whole).length) {
      const className = `.${cfg.prefix ? cfg.prefix + '-' : ''}rumi-${toHash(
        whole,
      )}`;
      classes[className] = stringify({[className]: whole});
    }

    return classes;
  };

  const css = (styles: TOrArrayT<ShortCircuitStyles>) => {
    const enabledStyles = (Array.isArray(styles) ? styles : [styles]).filter(
      (x) => x && x instanceof Object,
    );

    const transformedStyles = enabledStyles.map((x) =>
      transformSpecialProperties(x),
    );

    const mergedStyles = mergeStyles(transformedStyles);

    const classes = generateClassNames(mergedStyles);
    for (const [className, cssString] of Object.entries(classes)) {
      if (!isInStylesheet(className)) {
        addToStylesheet(className, cssString);
      }
    }

    const classNameGetter = () => {
      return Object.keys(classes)
        .map((x) => x.slice(1))
        .join(' ');
    };
    classNameGetter.styles = styles;
    return classNameGetter;
  };

  return {css, getCssText};
};
