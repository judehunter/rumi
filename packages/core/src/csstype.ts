import {StandardProperties, SimplePseudos} from 'csstype';
import {RumiConfig} from './createRumi';
import {TOrArrayT} from './utils';
// import * as Native from './css';

// export {Native};

export type CSSBaseStyles = StandardProperties<
  string | number,
  string | number
>;
// export interface CSSBaseStyles
//   extends Native.StandardLonghandProperties,
//     Native.StandardShorthandProperties,
//     Native.SvgProperties {}

export type CSSProperties = {
  [K in keyof CSSBaseStyles]:
    | CSSBaseStyles[K]
    | Array<Extract<CSSBaseStyles[K], string>>;
};

export type CSSPseudos<T extends RumiConfig> = {
  [K in SimplePseudos as `&${K}`]?: TOrArrayT<StylesObject<T>>;
};

export type CSSNested<T extends RumiConfig> = {
  [K in
    | `& *`
    | '& > *'
    | '& + *'
    | `& ~ *`
    | `${string}&${string}`
    | `${string}@${string}`]?: TOrArrayT<StylesObject<T>>;
};

export type CSSUtils<T extends RumiConfig> = {
  [Prop in keyof T['utils']]?: T['utils'][Prop] extends (...args: any) => any
    ? Parameters<T['utils'][Prop]>[0]
    : never;
};

export type CSSMedia<T extends RumiConfig> = {
  [Prop in keyof T['media']]?: StylesObject<T>;
};
export interface StylesObjectInterface<T extends RumiConfig>
  extends CSSProperties,
    CSSPseudos<T>,
    CSSNested<T> {}

export type StylesObject<T extends RumiConfig> =
  | (StylesObjectInterface<T> & CSSUtils<T>)
  | {styles: StylesObject<T>};

// export type StylesObject<T extends RumiConfig> =
//   | ({[Prop in keyof CSSBaseStyles]?: CSSBaseStyles[Prop]} & {
//       [Prop in keyof T['media']]?: StylesObject<T>;
// } & {
//   [Prop in keyof T['utils']]?: T['utils'][Prop] extends (
//     ...args: any
//   ) => any
//     ? Parameters<T['utils'][Prop]>[0]
//     : never;
// } & {
//       [K in Contains<'&'>]?: StylesObject<T>;
//     } & {
//       [K in keyof CSSNestedSelectors]?: StylesObject<T>;
//     })
//   | {styles: StylesObject<T>};
