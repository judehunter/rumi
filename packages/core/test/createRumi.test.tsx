import {createRumi} from '../src/createRumi';
import renderer from 'react-test-renderer';
import React from 'react';

describe('createRumi', () => {
  test('one class', () => {
    const {css} = createRumi({
      utils: {},
    });
    const cls1 = css({
      color: 'red',
    });
    const cls2 = css({
      color: 'red',
    });

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const tree = renderer
      .create(
        <>
          <div className={`${cls1()}`} />
          <div className={`${cls2()}`} />
        </>,
      )
      .toJSON()!;

    expect(cls1()).toEqual(cls2());
    expect(cls1().split(' ').length).toEqual(1);
    expect(tree).toMatchSnapshot();
  });

  test('atomic classes use rule name for hash', () => {
    const {css} = createRumi({
      utils: {},
    });

    const cls1 = css({
      color: 'red',
    });
    const cls2 = css({
      backgroundColor: 'red',
    });

    expect(cls1()).not.toEqual(cls2());
  });

  test('atomic and whole classes', () => {
    const {css} = createRumi({
      utils: {},
    });
    const cls1 = css({
      color: 'red',
      display: 'flex',
      '&:hover': {
        color: 'blue',
        display: 'block',
      },
      '&:focus': {
        color: 'violet',
      },
    });

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const tree = renderer.create(<div className={`${cls1()}`} />).toJSON()!;

    expect(cls1().split(' ').length).toEqual(3);
    expect(tree).toMatchSnapshot();
  });

  test('composition', () => {
    const {css} = createRumi({
      utils: {},
    });
    const cls1 = css({
      color: 'red',
      display: 'flex',
      '&:hover': {
        color: 'blue',
        display: 'block',
      },
      '&:focus': {
        color: 'violet',
      },
    });
  });
});
