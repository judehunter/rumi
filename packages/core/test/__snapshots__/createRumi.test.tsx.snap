// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRumi atomic and whole classes 1`] = `
<div
  className="rumi-gmqXFB rumi-dhzjXW rumi-hPVfiJ"
/>
`;

exports[`createRumi one class 1`] = `
Array [
  <div
    className="rumi-gmqXFB"
  />,
  <div
    className="rumi-gmqXFB"
  />,
]
`;
